### PR TITLE
Move profiling to DSE

### PR DESCRIPTION
**Pull request by FabioLuporini**
This PR makes a step forward towards Devito 2.0

The obsolete, hacky, and bug-prone profiler.py module is drastically trimmed, and its key functionality of "counting flops in the generated code" is moved to the DSE. It is indeed possible to count the number of operations by applying the DSE function ``estimate_cost``, which in turn wraps Sympy's  ``count_ops``, to any Sympy expression.

The operation of counting the memory accesses performed by a Sympy expression is also moved to the DSE (from Propagator).

Summing up, now the DSE returns a data structure containing:
- the transformed expressions
- a mapper to identify time-invariant expressions, out of the transformed expressions
- operations count
- memory accesses count

This PR depends on #170

### CODE REVIEW

File: devito/propagator.py [hunk 1]
```diff
@@ -261,89 +272,36 @@ def niters(self):
...
-            reads = set(flatten([e.rhs.atoms(Indexed) for e in expressions]))
-            writes = set(flatten([e.lhs.atoms(Indexed) for e in expressions]))
-            reads = set([access(s) for s in reads if filter(s)])
-            writes = set([access(s) for s in writes if filter(s)])
-            if mode == 'ideal':
-                return len(set(reads) | set(writes))
-            else:
-                return len(reads) + len(writes)
+    @property
+    def dataspace(self):
```
Comment (mlange05) [inline devito/propagator.py]: Docstring
